### ISSUE TITLE
Bilinear interpolation behavior inconsistent with TF and Caffe

### ISSUE THREAD

**Issue by libfun** [pytorch/pytorch#10604]
## Issue description

Trying to compare and transfer models between Caffe, TF and Pytorch found difference in output of bilinear interpolations between all. Caffe is using depthwise transposed convolutions instead of straightforward resize, so it's easy to reimplement both in TF and Pytorch.
However, there is difference between output for TF and Pytorch with `align_corners=False`, which is default for both.

## Code example

```Python
img = cv2.resize(cv2.imread('./lenna.png')[:, :, ::-1], (256, 256))
img = img.reshape(1, 256, 256, 3).astype('float32') / 255.
img = tf.convert_to_tensor(img)
output_size = [512, 512]
output = tf.image.resize_bilinear(img, output_size, align_corners=True)
with tf.Session() as sess:
        values = sess.run([output])
out_tf = values[0].astype('float32')[0]

img = img.reshape(1, 256, 256, 3).transpose(0, 3, 1, 2).astype('float32') / 255.
out_pt = nn.functional.interpolate(torch.from_numpy(nimg), 
                                   scale_factor=2, 
                                   mode='bilinear', 
                                   align_corners=True)
out_pt = out_pt.data.numpy().transpose(0, 2, 3, 1)[0]

print(np.max(np.abs(out_pt - out_tf)))  # output 5.6624413e-06
``` 
But
```Python
img = cv2.resize(cv2.imread('./lenna.png')[:, :, ::-1], (256, 256))
img = img.reshape(1, 256, 256, 3).astype('float32') / 255.
img = tf.convert_to_tensor(img)
output_size = [512, 512]
output = tf.image.resize_bilinear(img, output_size, align_corners=False)
with tf.Session() as sess:
        values = sess.run([output])
out_tf = values[0].astype('float32')[0]

img = img.reshape(1, 256, 256, 3).transpose(0, 3, 1, 2).astype('float32') / 255.
out_pt = nn.functional.interpolate(torch.from_numpy(nimg), 
                                   scale_factor=2, 
                                   mode='bilinear', 
                                   align_corners=False)
out_pt = out_pt.data.numpy().transpose(0, 2, 3, 1)[0]

print(np.max(np.abs(out_pt - out_tf)))  # output 0.22745097
``` 

Output diff:
![image](https://user-images.githubusercontent.com/301147/44241919-c4d2d180-a17b-11e8-81eb-9e44c5741d42.png)

Output of CoreML is consistent with TF, so it seems that there is a bug with implementation of bilinear interpolation with `align_corners=False` in Pytorch.

Diff is reproducible both on cpu and cuda with cudnn 7.1, cuda 9.1.

**Comment by SsnL** [comment]
Thanks for the report. We will look into this!

**Comment by libfun** [comment]
I've ran a few experiments and it seems that Pytorch with `align_corners=False` is consistent with python-opencv, PIL and skimage with `mode='edge'`. I'll look into CoreML closer today.

**Comment by ailzhang** [comment]
For pytorch implementation, here's the logic to compute src_idx <-> dst_idx mapping. 
https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/linear_upsampling.h#L22
For TF implementation, scale is caculated in the same way. https://github.com/tensorflow/tensorflow/blob/f66daa493e7383052b2b44def2933f61faf196e0/tensorflow/core/kernels/image_resizer_state.h#L41
But the src_idx <-> dst_idx mapping is different. https://github.com/tensorflow/tensorflow/blob/6795a8c3a3678fb805b6a8ba806af77ddfe61628/tensorflow/core/kernels/resize_bilinear_op.cc#L85 
I think Pytorch takes into account the pixel center when scaling, while TF doesn't.

**Comment by baowenbo** [comment]
> For pytorch implementation, here's the logic to compute src_idx <-> dst_idx mapping.
> https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/linear_upsampling.h#L22
> For TF implementation, scale is caculated in the same way. https://github.com/tensorflow/tensorflow/blob/f66daa493e7383052b2b44def2933f61faf196e0/tensorflow/core/kernels/image_resizer_state.h#L41
> But the src_idx <-> dst_idx mapping is different. https://github.com/tensorflow/tensorflow/blob/6795a8c3a3678fb805b6a8ba806af77ddfe61628/tensorflow/core/kernels/resize_bilinear_op.cc#L85
> I think Pytorch takes into account the pixel center when scaling, while TF doesn't.

So do you think there is any workaround in Pytorch to make it's interpolation consistent with that of TensorFlow?
I thought it may be possible to try `torch.nn.functional.grid_sample()` of Pytorch.

**Comment by AmitGarage** [comment]
> > For pytorch implementation, here's the logic to compute src_idx <-> dst_idx mapping.
> > https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/linear_upsampling.h#L22
> > For TF implementation, scale is caculated in the same way. https://github.com/tensorflow/tensorflow/blob/f66daa493e7383052b2b44def2933f61faf196e0/tensorflow/core/kernels/image_resizer_state.h#L41
> > But the src_idx <-> dst_idx mapping is different. https://github.com/tensorflow/tensorflow/blob/6795a8c3a3678fb805b6a8ba806af77ddfe61628/tensorflow/core/kernels/resize_bilinear_op.cc#L85
> > I think Pytorch takes into account the pixel center when scaling, while TF doesn't.
> 
> So do you think there is any workaround in Pytorch to make it's interpolation consistent with that of TensorFlow? I thought it may be possible to try `torch.nn.functional.grid_sample()` of Pytorch.

Using half_pixel_centers = True along with align_corners = False in tf.image.resize_bilinear will work same as torch.nn.functional.interpolate with align_corners = False

**Comment by TheDevilWillBeBee** [comment]
This is a PyTorch implementation of tf.compat.v1.image.resize_bilinear function.

`def symmetric_padding(im, padding):
    h, w = im.shape[-2:]
    left, right, top, bottom = padding

    x_idx = np.arange(-left, w + right)
    y_idx = np.arange(-top, h + bottom)

    def reflect(x, minx, maxx):
        """ Reflects an array around two points making a triangular waveform that ramps up
        and down,  allowing for pad lengths greater than the input length """
        rng = maxx - minx
        double_rng = 2 * rng
        mod = np.fmod(x - minx, double_rng)
        normed_mod = np.where(mod < 0, mod + double_rng, mod)
        out = np.where(normed_mod >= rng, double_rng - normed_mod, normed_mod) + minx
        return np.array(out, dtype=x.dtype)

    x_pad = reflect(x_idx, -0.5, w - 0.5)
    y_pad = reflect(y_idx, -0.5, h - 0.5)
    xx, yy = np.meshgrid(x_pad, y_pad)
    return im[..., yy, xx]

def tf_consistent_bilinear_upsample(imgs, scale_factor=1.0):
    b, c, h, w = imgs.shape
    assert h == w
    
    N = int(h * scale_factor)
    delta = (1.0 / N)
    xs = torch.linspace(-1.0 + 2 * delta, 1.0 - 2 * delta, N - 1)
    ys = torch.linspace(-1.0 + 2 * delta, 1.0 - 2 * delta, N - 1)
    grid = torch.meshgrid(xs, ys)
    gridy = grid[1]
    gridx = grid[0]
    gridx = symmetric_padding(gridx, (0, 1, 0, 1))
    gridy = symmetric_padding(gridy, (0, 1, 0, 1))
    grid = torch.stack([gridy, gridx], dim=-1).unsqueeze(0)
    output = torch.nn.functional.grid_sample(imgs, grid, mode='bilinear', padding_mode='zeros')
    
    return output
    `

**Comment by 980202006** [comment]
This is the correct solution, I tested the result on 1 dimension, the loss is less than 0.01.
